evennia/contrib/cooldowns.py

```diff
@@ -0,0 +1,188 @@
+
+    from evennia.contrib.cooldowns import Cooldownhandler
+    from evennia.utils.utils import lazy_property
+
+    @lazy_property
+    def cooldowns(self):
+        return CooldownHandler(self, db_attribute="cooldowns")
+
+Example:
+
+Assuming you've installed cooldowns on your Character typeclasses, you can
+use a cooldown to limit how often you can perform a command:
+
+class PowerAttack(Command):
+    def func(self):
+        if self.caller.cooldowns.ready("power attack"):
+            self.do_power_attack()
+            self.caller.cooldowns.set("power attack", 10)
+        else:
+            self.caller.msg("That's not ready yet!")
+"""
+
+import math
+import time
+
+
+class CooldownHandler:
+    """
+    Handler for cooldowns. This can be attached to any object that
+    supports DB attributes (like a Character or Account).
+
+    A cooldown is a timer that is usually used to limit how often
+    some action can be performed or some effect can trigger. When a
+    cooldown is first set, it counts down from the amount of time
+    provided back to zero, at which point it is considered ready again.
+
+    Cooldowns are named with an arbitrary string, and that string is used
+    to check on the progression of the cooldown. Each cooldown is tracked
+    separately and independently.
+
+    Cooldowns are saved persistently, so they survive reboots. This
+    module does not register or provide callback functionality for when
+    a cooldown becomes ready again. Users of cooldowns are expected to
+    query the state of any cooldowns they are interested in.
+
+    Methods:
+    - ready(name): Checks whether a given cooldown name is ready.
+    - time_left(name): Returns how much time is left on a cooldown.
+    - set(name, seconds): Sets a given cooldown to last for a certain
+        amount of time. Until then, ready() will return False for that
+        cooldown name.
+    - extend(name, seconds): Like set, but adds time to the given
+        cooldown name. If it doesn't exist yet, calling this is equivalent
+        to calling set.
+    - reset(cooldown): Resets a given cooldown, causing ready() to return
+        True for that cooldown immediately.
+    - clear(): Resets all cooldowns.
+    """
+
+    def __init__(self, obj, db_attribute="cooldowns"):
+        if not obj.attributes.has(db_attribute):
+            obj.attributes.add(db_attribute, {})
+
+        self.data = obj.attributes.get(db_attribute)
+        self.cleanup()
+
+    @property
+    def all(self):
+        """
+        Returns a list of all keys in this object.
+        """
+        return list(self.data.keys())
+
+    def ready(self, *args):
+        """
+        Checks whether all of the provided cooldowns are ready (expired).
+        If a requested cooldown does not exist, it is considered ready.
+
+        Args:
+            any (str): One or more cooldown names to check.
+        Returns:
+            (bool): True if each cooldown has expired or does not exist.
```

Comment:
A minor thing, but it matters nowadays when these docstrings are auto-parsed into documentation... 

- `any (str)` should be `*args (str)` these days. We use `any` in several other places, but this should be changed.
- `(bool):` should be just `bool:`.

Reply:
Thanks, fixed here and elsewhere in the PR.